driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +184,356 @@ public int getMaxWireVersion() {
-    }
+    /**
+     * This class maintains all resources that must be released in {@link AsyncCommandBatchCursor#close()}.
+     * It also implements a {@linkplain #doClose() deferred close action} such that it is totally ordered with other operations of
+     * {@link AsyncCommandBatchCursor} (methods {@link #tryStartOperation()}/{@link #endOperation()} must be used properly to enforce the order)
+     * despite the method {@link AsyncCommandBatchCursor#close()} being called concurrently with those operations.
+     * This total order induces the happens-before order.
+     * <p>
+     * The deferred close action does not violate externally observable idempotence of {@link AsyncCommandBatchCursor#close()},
+     * because {@link AsyncCommandBatchCursor#close()} is allowed to release resources "eventually".
+     * <p>
+     * Only methods explicitly documented as thread-safe are thread-safe,
+     * others are not and rely on the total order mentioned above.
+     */
+    @ThreadSafe
+    private final class ResourceManager {
+        private final Lock lock;
+        private volatile State state = State.IDLE;
+        @Nullable
+        private volatile AsyncConnectionSource connectionSource;
+        @Nullable
+        private volatile AsyncConnection pinnedConnection;
+        @Nullable
+        private volatile ServerCursor serverCursor;
+        private volatile boolean skipReleasingServerResourcesOnClose;
+
+        ResourceManager(final AsyncConnectionSource connectionSource,
+                @Nullable final AsyncConnection connectionToPin,
+                @Nullable final ServerCursor serverCursor) {
+            lock = new ReentrantLock();
+            if (serverCursor != null) {
+                this.connectionSource = notNull("connectionSource", connectionSource).retain();
+                if (connectionToPin != null) {
+                    this.pinnedConnection = connectionToPin.retain();
+                    connectionToPin.markAsPinned(Connection.PinningMode.CURSOR);
+                }
+            }
+            skipReleasingServerResourcesOnClose = false;
+            this.serverCursor = serverCursor;
+        }
 
-    private void endOperationInProgress() {
-        boolean closePending = withLock(lock, () -> {
-            isOperationInProgress = false;
-            return this.isClosePending;
-        });
-        if (closePending) {
-            close();
+        /**
+         * Thread-safe.
+         */
+        boolean operable() {
+            return state.operable();
         }
-    }
 
+        /**
+         * Thread-safe.
+         * Returns {@code true} iff started an operation.
+         * If {@linkplain State#operable() closed}, then returns false, otherwise completes abruptly.
+         *
+         * @throws IllegalStateException Iff another operation is in progress.
+         */
+        private boolean tryStartOperation() throws IllegalStateException {
```

Comment:
The reason `tryStartOperation`, `endOperation` are `private` is that they are not supposed to be called outside of `ResourceManager`. In `CommandBatchCursor`, these methods are an implementation detail of `ResourceManager.execute`. All "operations" are then clearly demarcated by this `execute` method, which is what I would expect to see in `AsyncCommandBatchCursor` if it were refactored to rely on `ResourceManager` similarly to `CommandBatchCursor`.

Naturally, `AsyncCommandBatchCursor.ResourceManager.execute` can't be implemented using the `try-finally` block, as it is done in `CommandBatchCursor`. Instead, an approach similar to `Locks.withLockAsync` (see https://github.com/mongodb/mongo-java-driver/pull/1134), or to `AsyncOperationHelper.executeRetryableReadAsync` (which is conceptually the same thing, with a different API and a bit different behavior) has to be used.

The current approach does not demarcate operations clearly and makes it difficult to see if each invocation of `tryStartOperation` is guaranteed to be followed by a corresponding `endOperation`. For example, `AsyncCommandBatchCursor.next` is supposed to be a single "operation" from the standpoint of `ResourceManager`, and it calls `tryStartOperation` once, but `endOperation` is called in many places. I would expect `endOperation` to be at least called every time we complete the callback passed to `AsyncCommandBatchCursor.next`, but this does not seem to be the case: for example, `getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback)` does not call `endOperation` if `t != null` (should this be fixed, or am I missing something?).